### ISSUE TITLE
java example project

### ISSUE THREAD

**Issue by jacksonlo** [carta/exscalabur#49]
example usage with java

**Comment by ksdsouza** [comment]
Created https://github.com/carta/exscalabur/pull/50 as an example of using the repo with Java

I think we'll need to do some work on making the API play nice with Java collections, cause right now it's kind of a pain

**Comment by jacksonlo** [comment]
What were you thinking in terms of the Java collections?

**Comment by ksdsouza** [comment]
There are two problems I saw with using Exscalabur out of the box with Java
* Need to use `JavaConverters` to convert from Java collections like `ArrayList`, to the function's `Iterable`/`Seq` param type
* Having to explicitly call `.apply()` for things like `DataRow`/`DataCell` etc.

Im not sure if there's a better way to address the first issue other than overloading for compatibility with the two collections hierarchies. Granted the java-support-overloads would just call the current methods after an implicit conversion, but still.

The second problem can easily be resolved through auxiliary constructors that do the `apply`'s current functionality.

Tldr; Right now, using Exscalabur in Scala seems seamless, but it feels more "hacky" in Java-land

**Comment by jacksonlo** [comment]
I think overload is a good idea and would keep the API concise on both ends (java/scala).

On the other hand, what do you think of making an entire Java class that wraps all the calls to the scala API so that can avoid the `.apply()` at the same time as using java converters and keeping that "compatibility class" isolated, and not touching the scala class

**Comment by ksdsouza** [comment]
I like the Java wrapper idea. We could separate the java-support logic by namespace by putting it in a separate package.

So scala users would still `import com.carta.exscalabur._`,  and if they need to work with java stuff they can use `import com.carta.java.exscalabur._`, where the stuff in the java package contains the wrappers of our current scala stuff

**Comment by jacksonlo** [comment]
yup makes sense to me!

**Comment by david-dufour-kw** [comment]
Would it be possible to just provide a package object with implicit conversions from Java to Scala types?

https://docs.scala-lang.org/tour/implicit-conversions.html

**Comment by ksdsouza** [comment]
I wrote a [dummy project](https://github.com/ksdsouza/TestProject) to try out package objects + implicit conversions and that didnt seem to help on the Java side (oddly enough, in a scala file using a java `List` will auto convert to the scala equivalent w/o needing an IC )

So i think a separate package for java users is the simplest soln.

**Comment by david-dufour-kw** [comment]
I see that the implicit conversion isn't necessary for the Scala test projedt from the dummy project but I don't see that it doesn't work for the Java test project. I guess it makes sense that you can't use something like `import scala.collection.JavaConverters._` inside of a Java project and you won't be able to convert it in the library because the parameters are the wrong type.

I think it's fine to go with the explicit wrapping approach.

**Comment by ksdsouza** [comment]
just to highlight,
[This part in the scala main file](https://github.com/ksdsouza/TestProject/blob/master/src/main/scala/com/tester/test/Main2.scala#L9) works just fine

[But this part in the java main file](https://github.com/ksdsouza/TestProject/blob/master/src/main/java/com/test/test2/Main.java#L12) doesn't make use of the IC, and will cause a compilation error

Will go ahead with the explicit wrapping approach in next PR

**Comment by ksdsouza** [comment]
This has the compatibility wrappers: https://github.com/carta/exscalabur/pull/51

Also, I don't have merge perms anymore, so you guys are gonna have to merge PRs in when they are approved.
